exercise/templates/exercise/exercise.html

```diff
@@ -79,28 +79,54 @@
 <script src="{% static 'exercise/duplicate_check.js' %}"></script>
 <script src="{% static 'exercise/chapter.js' %}"></script>
 <script
-    src="{% static 'exercise/language_link.js' %}"
-    data-set-lang="{{ instance|url:'set-enrollment-language' }}"></script>
+	src="{% static 'exercise/language_link.js' %}"
+	data-set-lang="{{ instance|url:'set-enrollment-language' }}"></script>
 <script>
 // Visualize loading in case of slow POST operations.
 /*$(function() {
-    var forms = $("#exercise-page-content").find("form:not([data-aplus-exercise] form)");
-    forms.on("submit", function(event) {
-        $("#submit-progress").removeClass("hide");
-    });
+	var forms = $("#exercise-page-content").find("form:not([data-aplus-exercise] form)");
+	forms.on("submit", function(event) {
+		$("#submit-progress").removeClass("hide");
+	});
 });*/
 
 // Add an Ajax exercise event listener to refresh the info column.
 window.addEventListener("message", function (event) {
-    if (event.data.type === "a-plus-refresh-stats") {
-        $("#submit-progress").addClass("hide");
-        var $stats = $("#exercise-info"),
-            url = $stats.data("url");
-        $stats.load(url, function() {
-            $stats.find('[data-toggle="tooltip"]').tooltip();
-        });
-    }
+	if (event.data.type === "a-plus-refresh-stats") {
+		$("#submit-progress").addClass("hide");
+		var $stats = $("#exercise-info"),
+			url = $stats.data("url");
+		$stats.load(url, function() {
+			$stats.find('[data-toggle="tooltip"]').tooltip();
+		});
+	}
 });
 </script>
+
+<script>
+var hash = window.location.hash
+
+//only register the events related to anchor links when theres a hash (e.g. #chapter-exercise-3)
+if (hash) {
+	//it isn't enough to just register a scroll event which disables setting the scroll
+	//location to the hash id. The scroll event fires every time the page loads, and
+	//whenever an exercise loads (as this changes the height of the page). To identify
+	//if a user manually scrolls we should receive 3 consecutive scroll events with no
+	//ajaxComplete events in between. An ajaxComplate event occuring signifies that the
+	//page scroll was caused by an exercise loading, and not the user.
+	var consecutiveScrollEvents = 0
+
+	window.addEventListener('scroll', () => {
+		consecutiveScrollEvents += 1
+	})
+
+	$( document ).ajaxComplete(() => {
+		if (consecutiveScrollEvents < 3) {
+			location.hash = hash
+			consecutiveScrollEvents = 0
+		}
+	})
```

Comment:
Could the `ajaxComplete` handler recognize that all exercises have been loaded and then remove the scroll event handler from the window? It feels unnecessary that the scroll event handler will always be called even long after the initial page load. Scroll events tend to happen a lot.

Reply:
I think a good way of removing the scroll listener would be to remove it when manual scrolling is detected. At that point the scroll listener is useless. 

```
function incrementScrollEvents() {
	consecutiveScrollEvents += 1;
}

window.addEventListener('scroll', incrementScrollEvents);

$( document ).ajaxComplete(() => {
	if (consecutiveScrollEvents < 3) {
		location.hash = hash;
		consecutiveScrollEvents = 0;
	}
	else {
		window.removeEventListener('scroll', incrementScrollEvents);
	}
})
```

Reply:
That sounds like a good idea! I would like you to implement it.